### PR TITLE
PaymentIntents Support

### PR DESCRIPTION
**Pull request by dweedon-stripe**
### Summary & motivation
Similar to `createToken`, `createSource` this PR adds element inference to the following:
- `stripe.createPaymentMethod`
- `stripe.handleCardPayment`

### Testing & documentation
- Wrote tests for `handleCardPayment` and `createPaymentMethod`
- Added PaymentIntents demos in a new demo file. 
- Updated documentation.

### CODE REVIEW

File: demo/paymentIntents/api.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      )
+    );
+  }
+
+  // WARNING
+  // DO NOT USE THE FOLLOWING CODE IN A PRODUCTION APPLICATION!
+  // Your secret key should NEVER be used in on your frontend.
+  // We are doing this here purely for demonstration reasons.
+  // In the real world, creating PaymentIntents needs to be done
+  // on your backend server.
```
Comment (asolove-stripe) [inline demo/paymentIntents/api.js]: We might want to make this even stronger and make clear the consequences: having your secret key on the frontend enables people to make charges on your behalf, which will be used by card testers, which will get your business banned from accepting credit card payments.

File: webpack.config.js [hunk 2]
```diff
@@ -26,6 +28,16 @@ const config = {
...
+    new HtmlWebpackPlugin({
+      inject: false,
+      filename: 'paymentIntents/index.html',
+      template: './demo/paymentIntents/index.html',
+    }),
+    new webpack.DefinePlugin({
+      'process.env.STRIPE_SECRET_KEY': JSON.stringify(
+        process.env.STRIPE_SECRET_KEY
+      ),
+    }),
```
Comment (asolove-stripe) [inline webpack.config.js]: I don't see any readme changes in this PR.

- It seems like we should at least add the new methods to the code block in the readme that starts `type FactoryProps = {`.
- Is it odd that the examples in the readme all use `createToken`, but this PR changes all demos to use `createPaymentMethod`? I wonder whether we should have a token demo, or a readme sample using cPM.

File: src/decls/Stripe.js [hunk 3]
```diff
@@ -25,4 +25,14 @@ declare type StripeShape = {
...
+  createPaymentMethod: (
+    type: string,
+    element: ElementShape,
+    data: mixed
+  ) => Promise<{paymentMethod?: MixedObject, error?: MixedObject}>,
+  handleCardPayment: (
+    clientSecret: string,
+    element: ElementShape | MixedObject,
+    options: mixed
+  ) => Promise<{paymentIntent?: MixedObject, error?: MixedObject}>,
```
Comment (asolove-stripe) [inline src/decls/Stripe.js]: Have we filed follow-up tasks to deal with `confirmPaymentIntent` and `handleCardAction`?

File: src/components/Provider.js [hunk 4]
```diff
@@ -51,7 +51,14 @@ const getOrCreateStripe = (apiKey: string, options: mixed): StripeShape => {
...
 
 const ensureStripeShape = (stripe: mixed): StripeShape => {
-  if (stripe && stripe.elements && stripe.createSource && stripe.createToken) {
+  if (
+    stripe &&
+    stripe.elements &&
+    stripe.createSource &&
+    stripe.createToken &&
+    stripe.createPaymentMethod &&
+    stripe.handleCardPayment
```
Comment (asolove-stripe) [inline src/components/Provider.js]: Hmm, this gets a little awkward. What if people pass in a Stripe instance that they instantiated themselves with one of the old PI beta flags that doesn't provide `createPaymentMethod`? Should we just say "this new major version only supports the unbeta'd state?" I think that's an ok decision, just want to check that it's deliberate.

File: webpack.config.js (hunk 2 above)
Comment (dweedon-stripe) [reply webpack.config.js]: Ah, I missed that all the demos switched over to cPM, will switch back. I wanted to move all the PI specific stuff into a demo.

Comment (dweedon-stripe) [reply webpack.config.js]: Also readme changes inc, still todo

File: src/decls/Stripe.js (hunk 3 above)
Comment (dweedon-stripe) [reply src/decls/Stripe.js]: `handleCardAction` will not require any work in RSE. We could all the types though. `confirmPaymentIntent` will be tricky, will add a follow up.

File: src/components/Provider.js (hunk 4 above)
Comment (dweedon-stripe) [reply src/components/Provider.js]: I think we have to make a major version cut that makes sense.

File: README.md [hunk 5]
```diff
@@ -191,22 +198,39 @@ class CheckoutForm extends React.Component {
...
+      .createPaymentMethod('card', data)
+      .then(({paymentMethod}) => {
+        console.log('Received Stripe PaymentMethod:', paymentMethod);
+      });
 
-    // However, this line of code will do the same thing:
-    //
-    // this.props.stripe.createToken({type: 'card', name: 'Jenny Rosen'});
+    // You can also use handleCardPayment with the Payment Intents API automatic confirmation flow
+    // documentation for more: https://stripe.com/docs/stripe-js/reference#stripe-handle-card-payment
```
[the remark this answers is not in the document]
Comment (dweedon-stripe) [reply README.md]: Ah yeah, it's supposed to be two sentences. Will fix.

**Comment by dweedon-stripe** [comment]
* Yes, did a manual test of 3DS
* Hmm, I'll add some more tests for hcp
* https://github.com/stripe/react-stripe-elements/pull/345/files#diff-2382a7144b80bd7c6cfe69ed37c02f0cR302

**Comment by dweedon-stripe** [comment]
Updated @asolove-stripe ptal

File: README.md [hunk 6]
```diff
@@ -212,19 +212,19 @@ class CheckoutForm extends React.Component {
...
-    // You can also use handleCardPayment with the Payment Intents API automatic confirmation flow
-    // documentation for more: https://stripe.com/docs/stripe-js/reference#stripe-handle-card-payment
+    // You can also use handleCardPayment with the Payment Intents API automatic confirmation flow.
+    // See more at: https://stripe.com/docs/stripe-js/reference#stripe-handle-card-payment
     this.props.stripe.handleCardPayment('{PAYMENT_INTENT_CLIENT_SECRET}', data);
 
-    // You can also use createToken to create tokens. See our tokens
-    // documentation for more: https://stripe.com/docs/stripe-js/reference#stripe-create-token
+    // You can also use createToken to create tokens. See our tokens.
+    // See more at: https://stripe.com/docs/stripe-js/reference#stripe-create-token
```
Comment (asolove-stripe) [inline README.md]: I think this and the next one now read wrong.

File: src/components/inject.js [hunk 7]
```diff
@@ -284,7 +288,7 @@ Please be sure the component that calls createSource or createToken is within an
       if (!clientSecret || typeof clientSecret !== 'string') {
         // If a bad value was passed in for clientSecret, throw an error.
         throw new Error(
-          `Invalid PaymentIntent clientSecret passed to handleCardPayment. Expected string, got ${typeof clientSecret}.`
+          `Invalid PaymentIntent Client Secret passed to handleCardPayment. Expected string, got ${typeof clientSecret}.`
```
Comment (asolove-stripe) [inline src/components/inject.js]: Hmm, I think this should not be capitalized?

File: src/components/inject.js [hunk 8]
```diff
@@ -317,7 +321,11 @@ Please be sure the component that calls createSource or createToken is within an
           return stripe.handleCardPayment(clientSecret, element);
         }
       } else {
-        if (!data || typeof data !== 'object') {
+        if (!data) {
+          throw new Error(
+            `Could not find a CardElement or CardNumberElement which which to perform handleCardPayment.`
+          );
+        } else if (typeof data !== 'object') {
```
Comment (asolove-stripe) [inline src/components/inject.js]: Are we sure object is strictly required? I thought you could pass a `clientSecret` and the id of a payment method or source as just a string?

Comment (dweedon-stripe) [reply src/components/inject.js]: That does not seem to work, and the logic inside Stripe.js does not appear to support that.

**Comment by dweedon-stripe** [comment]
changes... PTAL @asolove-stripe

**Comment by asolove-stripe** [comment]
lgtm, but there's a merge conflict 😢

**Pull request merged**